### PR TITLE
Fix propagating `only` and `exclude` to nested field instances

### LINKED ISSUE #1384: Dotted `only` and `exclude` not working for nested schema instances
#1229 fixed propagation of `only` and `exclude` for nested schema classes but no instances.
Dotted strings passed to `only` and `exclude` are not respected when a schema instance is nested.

```python
from marshmallow import Schema, fields


class Child(Schema):
    name = fields.String()
    age = fields.Integer()


class Parent(Schema):
    favorite = fields.Nested(Child, only=("name", "age"))
    children = fields.List(fields.Nested(Child, only=("name", "age")))


class Parent2(Schema):
    favorite = fields.Nested(Child(only=("name", "age")))
    children = fields.List(fields.Nested(Child(only=("name", "age"))))


family = {
    "favorite": {"name": "Lily", "age": 15},
    "children": [{"name": "Tommy", "age": 12}, {"name": "Lily", "age": 15}],
}
# should be the same
print(Parent(only=("favorite.age", "children.name")).dump(family))
print(Parent2(only=("favorite.age", "children.name")).dump(family))
```

Expected output:

```
{'children': [{'name': 'Tommy'}, {'name': 'Lily'}], 'favorite': {'age': 15}}
{'children': [{'name': 'Tommy'}, {'name': 'Lily'}], 'favorite': {'age': 15}}
```

Actual output:

```
{'favorite': {'age': 15}, 'children': [{'name': 'Tommy'}, {'name': 'Lily'}]}
{'favorite': {'age': 15, 'name': 'Lily'}, 'children': [{'age': 12, 'name': 'Tommy'}, {'age': 15, 'name': 'Lily'}]}
```

### PR DESCRIPTION
**Pull request by sloria**
Fix #1384

### CODE REVIEW

File: src/marshmallow/schema.py [hunk 1]
```diff
@@ -878,7 +875,9 @@ def __apply_nested_option(self, option_name, field_names, set_operation):
             setattr(self.declared_fields[key], option_name, new_options)
 
     def _init_fields(self):
-        """Update fields based on schema options."""
+        """Update self.fields, self.load_fields, and self.dump_fields based on schema options.
+        This method is private API.
```
Comment (sloria) [inline src/marshmallow/schema.py:879]: This was always private even though we don't actually use `__`, so I don't consider this a breaking change.

**Comment by sloria** [comment]
Merging this because I don't want this bug and the bug in #1160 in the wild for too long.

File: src/marshmallow/fields.py [hunk 2]
```diff
@@ -468,8 +468,17 @@ def schema(self):
             # Inherit context from parent.
             context = getattr(self.parent, "context", {})
             if isinstance(self.nested, SchemaABC):
-                self._schema = self.nested
+                self._schema = copy.deepcopy(self.nested)
```
Comment (metheoryt) [inline src/marshmallow/fields.py:471]: I have nested schema instance which context contains non-pickling object. In my case it is `Crypto.Cipher.AES` object which i tried to deepcopy in console and got an error. Should I start an issue?

**Comment by metheoryt** [comment]
Hi guys! I've leaved a comment on a diff page, in general now I cannot use non-deepcopyable objects in nested schema context. I use `Crypto.Cipher.AES` from `pycryptodome` package

File: src/marshmallow/fields.py (hunk 2 above)
Comment (sloria) [reply src/marshmallow/fields.py:471]: Does it fix it if you use `copy.copy` instead?

```suggestion
                self._schema = copy.copy(self.nested)
```

Now that I think of it, I'm not sure we even need to deepcopy here. Let me know if the above fixes your issue.

Comment (metheoryt) [reply src/marshmallow/fields.py:471]: @sloria yep, copy works just fine! waiting for release, thanks!

Comment (sloria) [reply src/marshmallow/fields.py:471]: Thanks for checking. The fix is released in 3.1.1.

**Pull request merged**